src/clp_ffi_py/ir/Query.hpp

```diff
@@ -0,0 +1,200 @@
+            std::numeric_limits<ffi::epoch_time_ms_t>::max()};
+
+    /**
+     * Ideally, when searching an IR stream with a query, the search terminates
+     * once the timestamp exceeds the search upper bound. However, the timestamp
+     * might not be monotonically increasing in a CLP IR stream. It can be
+     * locally disordered due to thread contention. To safely stop searching, we
+     * need to ensure that the current timestamp has exceeded the search's upper
+     * bound by a reasonable margin. This margin defaults to the following
+     * constant. During the query initialization, it will be applied to the
+     * upper bound timestamp to generate a timestamp which indicates it is safe
+     * to terminate the search.
+     */
+    static constexpr ffi::epoch_time_ms_t const cDefaultSearchTerminationMargin{
+            static_cast<ffi::epoch_time_ms_t>(60 * 1000)};
+
+    /**
+     * Constructs an empty query object that will match all logs. The wildcard
+     * list is empty and the timestamp range is set to include all the valid
+     * Unix epoch timestamps.
+     */
+    explicit Query()
+            : m_lower_bound_ts{cTimestampMin},
+              m_upper_bound_ts{cTimestampMax},
+              m_search_termination_ts{cTimestampMax} {};
+
+    /**
+     * Constructs a new query object with the given timestamp range with an
+     * empty wildcard list.
+     * @param search_time_lower_bound Start of search time range (inclusive).
+     * @param search_time_upper_bound End of search time range (inclusive).
+     * @param search_termination_margin The margin used to determine the search
+     * termination timestamp.
+     */
+    explicit Query(
+            ffi::epoch_time_ms_t search_time_lower_bound,
+            ffi::epoch_time_ms_t search_time_upper_bound,
+            ffi::epoch_time_ms_t search_termination_margin = cDefaultSearchTerminationMargin
+    )
+            : m_lower_bound_ts{search_time_lower_bound},
+              m_upper_bound_ts{search_time_upper_bound},
+              m_search_termination_ts{
+                      (cTimestampMax - search_termination_margin > search_time_upper_bound)
+                              ? search_time_upper_bound + search_termination_margin
+                              : cTimestampMax} {
+        throw_if_ts_range_invalid();
+    }
+
+    /**
+     * Constructs a new query object with the given timestamp range and a
+     * wildcard list.
+     * @param search_time_lower_bound Start of search time range (inclusive).
+     * @param search_time_upper_bound End of search time range (inclusive).
+     * @param wildcard_list A reference to the wildcard queries vector, whose
+     * data will be transferred using std::move to initialize m_wildcard_list.
+     * @param search_termination_margin The margin used to determine the search
+     * termination timestamp. By default, this value is set to
+     * cDefaultSearchTerminationMargin.
```

Comment:
Don't think you need to specify the default value since any decent doc site generator will pick up the default value from the code.